sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvRowConversions.java

```diff
@@ -0,0 +1,166 @@
+      return new AutoValue_CsvRowConversions_RowToCsv.Builder();
+    }
+
+    /** The expected {@link Schema} of the {@link Row} input. */
+    abstract Schema getSchema();
+
+    /** The {@link CSVFormat} of the converted {@link Row} input. */
+    abstract CSVFormat getCSVFormat();
+
+    /** The order and subset of {@link Schema} field names to drive the {@link Row} conversion. */
+    abstract List<String> getSchemaFields();
+
+    /** Builds a header from {@link #getSchemaFields()} using the {@link #getCSVFormat()}. */
+    String buildHeader() {
+      return buildHeaderFrom(getSchemaFields(), getCSVFormat());
+    }
+
+    /**
+     * Converts a {@link Row} to a CSV string formatted using {@link #getCSVFormat()}. Null values
+     * are converted to empty strings.
+     */
+    @Override
+    public String apply(Row input) {
+      List<String> schemaFields = getSchemaFields();
+      Object[] values = new Object[schemaFields.size()];
+      for (int i = 0; i < schemaFields.size(); i++) {
+        String name = schemaFields.get(i);
+        values[i] = input.getValue(name);
+      }
+      return getCSVFormat().format(values);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder {
+
+      /** The expected {@link Schema} of the {@link Row} input. */
+      abstract Builder setSchema(Schema schema);
+
+      abstract Schema getSchema();
+
+      /** The {@link CSVFormat} of the converted {@link Row} input. */
+      abstract Builder setCSVFormat(CSVFormat format);
+
+      /** The order and subset of {@link Schema} field names to drive the {@link Row} conversion. */
+      abstract Builder setSchemaFields(List<String> fields);
+
+      abstract List<String> getSchemaFields();
+
+      abstract RowToCsv autoBuild();
+
+      final RowToCsv build() {
+
+        validateHeaderAgainstSchema(getSchemaFields(), getSchema());
+
+        return autoBuild();
+      }
+    }
+  }
+
+  /** Formats columns into a header String based on {@link CSVFormat}. */
+  static String buildHeaderFrom(List<String> columns, CSVFormat csvFormat) {
+    StringBuilder builder = new StringBuilder();
+    try {
+      boolean newRecord = true;
+      for (String name : columns) {
+        csvFormat.print(name, builder, newRecord);
+        newRecord = false;
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+    return builder.toString();
+  }
+
+  private static void validateHeaderAgainstSchema(List<String> schemaFields, Schema schema) {
+    if (schema.getFieldCount() == 0) {
+      throw new IllegalArgumentException("schema is empty");
+    }
+    if (schemaFields.isEmpty()) {
+      throw new IllegalArgumentException(
+          "Columns is empty. An intent to not override columns, should assign null to the columns parameter.");
+    }
+    List<String> missing = new ArrayList<>();
+    List<Field> invalid = new ArrayList<>();
+
+    for (String name : schemaFields) {
+      if (name.isEmpty()) {
```

Comment:
A quick look seems to show that we can have schemas and CSVs with empty field names

I still think it would make sense to ensure that `name != null`